### PR TITLE
Add autostart option to BusABC.send_periodic() to fix issue #1848

### PR DESCRIPTION
**Pull request by SWolfSchunk**
I have fixed my own issue #1848

### CODE REVIEW

**Review by hardbyte — APPROVED** [review]
Seems reasonable

**Comment by SWolfSchunk** [comment]
It seems like the test "testProducerConsumer" fails randomly on random pypy versions (3.9, 3.10) on random platforms (windows here so far, but other requests show the same behavior on ubuntu and macos).
I can not appraise if it is really a problem depending on the changes of other requests but here i am sure it is not coming from the changes I made. Maybe I should open a bug report fort that but maybe this is a problem of the threading implementation of pypy?

**Comment by zariiii9003** [comment]
The autostart parameter is ignored by socketcan and ixxat, right? This should either be fixed or documented.
And yes, the PyPy tests are flaky, it's probably not related to your changes.

File: can/interfaces/socketcan/socketcan.py [hunk 1]
```diff
@@ -823,13 +824,16 @@ def _send_periodic_internal(
         :class:`CyclicSendTask` within BCM provides flexibility to schedule
         CAN messages sending with the same CAN ID, but different CAN data.
 
-        :param messages:
+        :param msgs:
```
Comment (grant-allan-ctct) [inline can/interfaces/socketcan/socketcan.py:827]: 👍

**Comment by zariiii9003** [comment]
> No that's not right the autostart parameter is not ignored by neither socketcan nor ixxat. 

Your change only affects the thread-based fallback. However, the socketcan broadcast manager start immediately.

**Comment by SWolfSchunk** [comment]
Hopefully I have fixed the missing parts now. Thanks for your hint @zariiii9003 .
I am still not sure about the changes for socketcan devices.
From how i read the code of its CyclicSendTask.__init__() instead of other devices they where not autostarting before why i set the default value to autostart=False in this case for backward compatibility.

File: can/interfaces/socketcan/socketcan.py [hunk 2]
```diff
@@ -376,7 +380,9 @@ def _tx_setup(
         for message in messages:
             body += build_can_frame(message)
         log.debug("Sending BCM command")
-        send_bcm(self.bcm_socket, header + body)
+
+        if send:
```
Comment (zariiii9003) [inline can/interfaces/socketcan/socketcan.py:384]: I think this check should be in some other place, but the whole class needs some refactoring. I'll fix this later.

File: can/interfaces/socketcan/socketcan.py [hunk 3]
```diff
@@ -823,13 +830,16 @@ def _send_periodic_internal(
...
+        :param msgs:
             The message(s) to be sent periodically.
         :param period:
             The rate in seconds at which to send the messages.
         :param duration:
             Approximate duration in seconds to continue sending messages. If
             no duration is provided, the task will continue indefinitely.
+        :param autostart:
+            If True (the default) the sending task will immediately start after creation.
+            Otherwise, the task has to be started by calling start() on it.
```
Comment (zariiii9003) [inline can/interfaces/socketcan/socketcan.py:842]: Same as above, `start()` should be linked.

**Review by zariiii9003 — CHANGES_REQUESTED** [review]

File: can/interfaces/socketcan/socketcan.py [hunk 4]
```diff
@@ -349,10 +350,13 @@ def __init__(
...
         self.task_id = task_id
-        self._tx_setup(self.messages)
+        self._tx_setup(self.messages, send=autostart)
 
     def _tx_setup(
-        self, messages: Sequence[Message], raise_if_task_exists: bool = True
+        self,
+        messages: Sequence[Message],
+        raise_if_task_exists: bool = True,
+        send: bool = True,
```
[the remark this answers is not in the document]
Comment (SWolfSchunk) [reply can/interfaces/socketcan/socketcan.py:359]: To make everything clearer I have had a closer look at the code and decided to guard the whole  call of self._tx_setup() with if autostart to remove the send parameter as the name autostart would not really fit.
I also think that the name of the _tx_setup() should be changed to make clearer what it really does.

File: can/interfaces/socketcan/socketcan.py (hunk 2 above)
Comment (SWolfSchunk) [reply can/interfaces/socketcan/socketcan.py:384]: Se the comment above

**Review by zariiii9003 — APPROVED** [review]

**Comment by zariiii9003** [comment]
Thank you 👍

**Pull request merged**